### PR TITLE
Remove database consistency validation of clinical attributes

### PR DESCRIPTION
**Pull request by fedde-s**
Clinical attributes used to share a single namespace across all studies, meaning that once the meaning of an attribute name was defined (with a display name, mouseover description, data type, sample or patient level and priority), all subsequent studies would have to adhere to that. Commit b6dcfc082116293178b0cd762b0be7aba3152d51 made clinical attributes study-specific. Because of that, validating whether the definitions of attributes in a study are consistent with those configured in the portal is now no longer applicable. Removing the dependency on these APIs fixes the problems caused by duplicate entries introduced into this API's output, reported in issue #1804.

Changes proposed in this pull request:
- Remove database consistency validation for clinical attributes
- Remove clinical attribute API calls from the validator
- Clarify the message emitted when attributes do break requirements by adding the line and column numbers of the offending value

### CODE REVIEW

**Comment by fedde-s** [comment]
@jjgao: This is the fix Pieter asked about in the **etl** channel on Slack. It might be easiest to merge pull request #1809 to `hotfix` first, as these are the changes that have already been approved as part of pull request #1567. After that, the only additional changes to be reviewed in this pull request will be the three commits that fix issue #1804.

**Comment by fedde-s** [comment]
I will replace the first five commits (the ones in #1809) by the original commits that are already in rc, to prevent merging issues.

**Comment by fedde-s** [comment]
Done.

File: core/src/main/scripts/importer/validateData.py [hunk 1]
```diff
@@ -1284,24 +1283,146 @@ class ClinicalValidator(Validator):
...
     NULL_VALUES = ["[not applicable]", "[not available]", "[pending]", "[discrepancy]","[completed]","[null]", "", "na"]
     ALLOW_BLANKS = True
+    METADATA_LINES = ('display_name',
+                      'description',
+                      'datatype',
+                      'priority')
+
+    # attributes required to have certain properties because of hard-coded use
+    # TODO add unit tests for this functionality
+    PREDEFINED_ATTRIBUTES = {
```
Comment (pieterlukasse) [inline core/src/main/scripts/importer/validateData.py]: this is repeating what is already in #1809

Comment (fedde-s) [inline core/src/main/scripts/importer/validateData.py]: You are looking at the exact same commit.

File: core/src/main/scripts/importer/validateData.py [hunk 2]
```diff
@@ -1569,7 +1564,7 @@ def checkHeader(self, cols):
                                 extra={'line_number': self.line_number,
                                        'column_number': col_index + 1,
                                        'cause': col_name})
-                    else:
+                    elif not self.relaxed_mode:
```
Comment (pieterlukasse) [inline core/src/main/scripts/importer/validateData.py]: please add some comments here for why this is not applicable to relaxed_mode

File: core/src/main/scripts/importer/validateData.py [hunk 3]
```diff
@@ -1640,23 +1593,15 @@ def checkLine(self, data):
...
-            
+
             # if not blank, check if values match the datatype
             if value.strip().lower() in self.NULL_VALUES:
                 pass
             elif data_type == 'NUMBER':
                 if not self.checkFloat(value):
                     self.logger.error(
-                        'Value of attribute to be loaded as NUMBER is not a real number' + according_to_portal,
+                        'Value of attribute to be loaded as NUMBER is not a real number',
```
Comment (pieterlukasse) [inline core/src/main/scripts/importer/validateData.py]: change message to  'Value of attribute is not a real number'

File: core/src/main/scripts/importer/validateData.py [hunk 4]
```diff
@@ -1770,11 +1715,11 @@ def checkHeader(self, cols):
                        'column_number': self.cols.index('SAMPLE_ID'),
                        'cause': 'SAMPLE_ID'})
         # refuse to define attributes also defined in the sample-level file
-        for new_attribute in self.newly_defined_attributes:
+        for new_attribute in self.defined_attributes:
```
Comment (pieterlukasse) [inline core/src/main/scripts/importer/validateData.py]: `for defined_attribute in self.defined_attributes`?

File: core/src/test/scripts/unit_tests_validate_data.py [hunk 5]
```diff
@@ -1298,26 +1283,27 @@ def test_unnecessary_and_wrong_stable_id(self):
...
-        # for missing survival status attributes, 4 for patients with no samples
-        # and 1 for not being able to parse the header in the clinical file
-        self.assertEqual(len(record_list), 10)
+        # we expect 9 errors or warnings:
+        # 1 for not being able to parse the header in the clinical file,
+        # 1 for the sample attribute CANCER_TYPE in a patient-level file,
+        # 1 for CANCER_TYPE being in DEFINED_SAMPLE_ATTRIBUTES
+        # 2 for missing survival status attributes and
+        # 4 for patients with no samples
+        self.assertEqual(len(record_list), 9)
```
Comment (pieterlukasse) [inline core/src/test/scripts/unit_tests_validate_data.py]: this is better implemented using the error file diff at the end

**Comment by pieterlukasse** [comment]
Look very good :+1:  Just some minor comments.

File: core/src/test/scripts/unit_tests_validate_data.py (hunk 5 above)
Comment (fedde-s) [inline core/src/test/scripts/unit_tests_validate_data.py]: I agree. But the design of this test case is not the issue I intended to fix in this pull request.

**Comment by fedde-s** [comment]
Thank you for the review, @pieterlukasse. I addressed most of your comments, Although the last two, I'd argue are a separate (though minor) issue introduced in pull request #1581.

**Comment by fedde-s** [comment]
...and now I fixed the test that started failing because one of the amendments changed an output message.

**Comment by fedde-s** [comment]
As @pieterlukasse requested, I added three commits to clarify the purpose of a few automated tests I edited and adjust their assertions to better match this purpose.

File: core/src/test/scripts/system_tests_validate_data.py [hunk 6]
```diff
@@ -47,24 +39,24 @@ def tearDown(self):
...
-        except WindowsError:
-            # ignore this Windows specific error...probably happens because of virus scanners scanning the temp file...
-            pass        
+            self.assertTrue(os.path.exists(tmp_file_name))
+            with open(tmp_file_name, 'rU') as out_file, \
+                    open(expected_file_name, 'rU') as ref_file:
+                # generate a diff between the expected and the generated file,
+                # even if the diff is longer than a few lines
+                self.maxDiff = None
+                self.assertMultiLineEqual(ref_file.read(), out_file.read())
```
Comment (pieterlukasse) [inline core/src/test/scripts/system_tests_validate_data.py]: works :+1:

**Pull request merged**